store/store.go

```diff
@@ -1178,6 +1198,49 @@ var download = func(name, downloadURL string, user *auth.UserState, s *Store, w
 	return err
 }
 
+// downloadDelta downloads the delta for the preferred format, returning the path.
+func (s *Store) downloadDelta(name string, downloadDir string, downloadInfo *snap.DownloadInfo, pbar progress.Meter, user *auth.UserState) (string, error) {
+
+	if len(downloadInfo.Deltas) != 1 {
+		return "", errors.New("store returned more than one delta")
+	}
+
+	deltaInfo := downloadInfo.Deltas[0]
+
+	if deltaInfo.Format != s.deltaFormat {
+		return "", fmt.Errorf("store returned delta with format %q", deltaInfo.Format)
+	}
+
+	deltaName := fmt.Sprintf("%s_%d_%d_delta.%s", name, deltaInfo.FromRevision, deltaInfo.ToRevision, deltaInfo.Format)
+
+	w, err := os.Create(path.Join(downloadDir, deltaName))
+	if err != nil {
+		return "", err
+	}
+	deltaPath := w.Name()
+	defer func() {
+		if cerr := w.Close(); cerr != nil && err == nil {
+			err = cerr
+		}
+		if err != nil {
+			os.Remove(w.Name())
+			deltaPath = ""
+		}
+	}()
+
+	url := deltaInfo.AnonDownloadURL
+	if url == "" || user != nil {
+		url = deltaInfo.DownloadURL
+	}
+
+	err = download(deltaName, url, user, s, w, pbar)
+	if err != nil {
+		return "", err
+	}
+
```

Comment:
We chatted about this - and you mentioned that it is fine without a w.Sync() even, as xdelta doesn't go through raw io, but goes through the vfs.